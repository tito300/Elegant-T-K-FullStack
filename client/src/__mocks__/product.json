{
    "discreption": "random description",
    "id": "1234",
    "lastModified": "2019-01-10T05:19:59.000Z",
    "material": "fiber",
    "photo": "imgs/1.jpg",
    "price": "20.00",
    "primaryColor": "red",
    "purchaseCount": 6,
    "rating": 3,
    "secondaryPhotos": [
        {
            "color": "black",
            "link": "imgs/1-1.jpg"
        },
        {
            "color": "grey",
            "link": "imgs/1-2.jpg"
        }
    ],
    "title": "random product",
    "uploadDate": "2018-08-07T20:37:06.682Z",
    "variants": {
        "male": [
            {
                "color": "red",
                "sizes": [
                    {
                        "color": "red",
                        "gender": "male",
                        "photo": "imgs/1.jpg",
                        "price": "23.00",
                        "size": "l"
                    },
                    {
                        "color": "red",
                        "gender": "male",
                        "photo": "imgs/1.jpg",
                        "price": "23.00",
                        "size": "s"
                    }
                ]
            }
        ],
        "female": [
            {
                "color": "black",
                "sizes": [
                    {
                        "color": "black",
                        "gender": "female",
                        "photo": "imgs/1.jpg",
                        "price": "21.00",
                        "size": "l"
                    },
                    {
                        "color": "black",
                        "gender": "female",
                        "photo": "imgs/1.jpg",
                        "price": "231.00",
                        "size": "s"
                    }
                ]
            }
        ]
    },
    "weight": "5"
}