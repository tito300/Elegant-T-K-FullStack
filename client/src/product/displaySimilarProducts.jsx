import React from 'react';

const DisplaySimilarProducts = ({ product }) => {
    return (
        <>
            <p className='similarItems'>Similar Products</p>
            <div className='bottomGallery'>
                <div>
                    <i className="fa fa-angle-left" aria-hidden="true"></i>
                </div>
                <div className="bottomPhotosGallery">
                    <div className="bottomItem">
                        <img className='galleryPhoto' src={`/${product.photo}`} />
                    </div>
                    <div className="bottomItem">
                        <img className='galleryPhoto' src={`/${product.photo}`} />
                    </div>
                    <div className="bottomItem">
                        <img className='galleryPhoto' src={`/${product.photo}`} />
                    </div>
                    <div className="bottomItem">
                        <img className='galleryPhoto' src={`/${product.photo}`} />
                    </div>
                    <div className="bottomItem">
                        <img className='galleryPhoto' src={`/${product.photo}`} />
                    </div>
                </div>
                <div>
                    <i className="fa fa-angle-right" aria-hidden="true"></i>
                </div>
            </div>
        </>
    )
}

export default DisplaySimilarProducts;