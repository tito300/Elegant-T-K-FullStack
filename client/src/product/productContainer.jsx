import React, { Component } from 'react';
import axios from 'axios';
import Product from './product.jsx';
import ErrorBoundary from './errorBoundaries/productError';


export class ProductContainer extends Component {
    state = {
        product: null,
        pending: true,
        err: false,
        errMsg: null,
    }

    container = {
        width: '1200px',
        margin: '150px auto 0',
        paddingTop: '30px',
        overFlow: 'hidden',
        minHeight: 'calc(100vh - 78px)',
    }

    faSpinner = {
        width: 'fit-content',
        margin: '50px auto 150px auto',
        color: 'gray',
    }

    componentDidMount() {
        // debugger;
        let { computedMatch } = this.props;
        axios.get(`/api/products/product/${computedMatch.params.id}`)
            .then(({ data }) => {
                this.setState({ product: data, pending: false })
            })
            .catch((err) => {
                console.log('catch called');
                let msg = '';
                if (err.response.status === 404) {
                    msg = '404: product is not available. Please contact us or come back later.'
                } else {
                    msg = 'server is not responding, please try refreshing the page or contact us.'
                }
                this.setState({ pending: false, err: true, errMsg: msg });
            })
    }

    componentDidUpdate() {
        let { computedMatch } = this.props;
        debugger;
        if (computedMatch.params.id !== this.state.product.id) {
            axios.get(`/api/products/product/${computedMatch.params.id}`)
                .then(({ data }) => {
                    this.setState({ product: data, pending: false })
                })
                .catch((err) => {
                    console.log('catch called');
                    let msg = '';
                    if (err.response.status === 404) {
                        msg = '404: product is not available. Please contact us or come back later.'
                    } else {
                        msg = 'server is not responding, please try refreshing the page or contact us.'
                    }
                    this.setState({ pending: false, err: true, errMsg: msg });
                })
        }
    }

    render() {
        if (!this.state.err) {
            return (
                <div className="product-display" style={this.container}>
                    {this.state.product ?
                        <ErrorBoundary>
                            <Product product={this.state.product} />
                        </ErrorBoundary>
                        : <div style={this.faSpinner} >
                            <i className="fa fa-spinner fa-pulse fa-3x fa-fw"></i>
                        </div>
                    }
                </div>
            )
        }
        if (this.state.err) {
            return (<div className="product-list__Error">
                <p className="product-list__Error_title" >
                    Sorry!
                </p>
                <p className="product-list__Error_details">
                    {this.state.errMsg}
                </p>
            </div>)
        }
    }
}

export default ProductContainer;
