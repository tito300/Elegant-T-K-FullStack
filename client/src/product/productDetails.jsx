import React from 'react';
import { Link } from 'react-router-dom'
import AddToCartButton from '../common/addToCart-button.jsx';
import propType from 'prop-types';


let jsUcfirst = function (string) {
    if (typeof string === 'string') {
        return string.charAt(0).toUpperCase() + string.slice(1);
    } else {
        console.error('variable string is not a string. it is: ' + JSON.stringify(string));
        throw new Error('title is not a string at jsUcfirst');
    }
}

let ProductDetails = (props) => {

    let { success, failed, addItemToCartSuccess, color, size, gender, product } = props;

    let brand = (product.brand && product.brand !== "") ? product.brand : 'Not available';
    return (
        <>
            <h2 className='title' >{jsUcfirst(product.title)}</h2>
            <p className='brand' >{`brand: `}
                <span>{brand}</span>
            </p>
            <div className='priceContainer' >
                <span className='price' >Price: <span className='priceNumber' >${product.price}</span></span>
                {(!success && !failed) ?
                    <AddToCartButton
                        item={product}
                        addItemToCartSuccess={addItemToCartSuccess}
                        options={{ color, size, gender }}
                    />
                    : (!success && failed) ?
                        <div style={{ width: '108px' }}>
                            <AddToCartButton
                                item={product}
                                addItemToCartSuccess={addItemToCartSuccess}
                                options={{ color, size, gender }}
                            />
                            <p className='addToCartWarning' style={{
                                position: 'relative'
                            }}>
                                {'Make sure to select\n color'}
                            </p>
                        </div>

                        : <div>
                            <svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 130.2 130.2">
                                <circle className="path circle" fill="none" stroke="#73AF55" strokeWidth="6" strokeMiterlimit="10" cx="65.1" cy="65.1" r="62.1" />
                                <polyline className="path check" fill="none" stroke="#73AF55" strokeWidth="6" strokeLinecap="round" strokeMiterlimit="10" points="100.2,40.2 51.5,88.8 29.8,67.5 " />
                            </svg>
                            <Link to='/cart/main'><p className="success">View Cart</p></Link>
                        </div>
                }
            </div>
            <div className='discreptionContainer' >
                <p className='discreptionText' >{product.discreption}</p>
            </div>

        </>)

}

ProductDetails.prototype = {
    success: propType.bool,
    failed: propType.bool,
    addItemToCartSuccess: propType.func,
    color: propType.string,
    size: propType.string,
    gender: propType.string,
    product: propType.object,
}

export default ProductDetails;