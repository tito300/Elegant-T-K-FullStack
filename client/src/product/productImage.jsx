import React, { Component } from 'react';

const ProductImage = ({ product, activePicture, handleSelectPhoto }) => {
    return (<div className='photoSection' >
        <div className='photoContainer' >
            <img src={`/${activePicture}`} className='photo' alt="" />
        </div>
        <div className='gallerySection' >
            <div>
                <i className="fa fa-angle-left" aria-hidden="true"></i>
            </div>
            <div className='galleryContainer' >
                {product.secondaryPhotos.length !== 0 ? product.secondaryPhotos.map((photo, i) => {
                    return (<div className='imgContainer' key={i} >
                        <img className='galleryPhoto' src={`/${photo}`} photoindex={i} onClick={handleSelectPhoto} alt='product' />
                    </div>)
                })
                    : (<div className='imgContainer' >
                        <img className='galleryPhoto' src={`/${activePicture}`} alt='product' />
                    </div>)}
            </div>
            <div>
                <i className="fa fa-angle-right" aria-hidden="true"></i>
            </div>
        </div>
    </div>);
}

export default ProductImage;