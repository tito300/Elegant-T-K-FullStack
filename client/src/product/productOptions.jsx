import React from 'react';

export default ({ variants, handleOptions, gender, color, size }) => {
    if (!variants.male.length && !variants.female.length) {
        return (
            <div className='optionsContainer'>
                <p style={{ color: 'red' }}>Sorry, this item is out of stock</p>
            </div>
        )
    }

    return (
        <div className='optionsContainer'>
            <div>
                <p className="optionsTitle">Gender</p>
                <select className="selectGender" onChange={handleOptions}>
                    <option value='male' onChange={handleOptions}>Male</option>
                    <option value='female' onChange={handleOptions}>Female</option>
                    <option value="">none</option>
                </select>
            </div>
            <div>
                <p className="optionsTitle">Color</p>
                <div className="colorsContainer">
                    {variants[gender].map(size => {
                        return (<span
                            id={size.color}
                            className={color === size.color ? 'colorOption selectedColor' : 'colorOption'}
                            title={`color: ${size.color}`}
                            onClick={handleOptions} />)
                    }
                    )}
                </div>
            </div>
            {color !== null && gender ?
                <div>
                    <p className="optionsTitle">Size</p>
                    <select className="selectSize" value={size} onChange={handleOptions}>
                        {mapSizeOptions(variants, gender, color)}
                    </select>
                </div> : null}
        </div>
    )
}

/**
 * finds available sizes for a given color and gender and maps them to html option tags
 * @param {*} variants 
 * @param {*} gender 
 * @param {*} color 
 */
let mapSizeOptions = (variants, gender, color) => {
    return variants[gender].find(e => e.color === color).sizes.map(size => {
        return (
            <option key={size.id} value={size.size} >{size.size}</option>
        )
    })
}