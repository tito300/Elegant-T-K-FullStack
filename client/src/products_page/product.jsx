import React, { Component } from 'react';
import { throws } from 'assert';
import propType from 'prop-types';
import ProductImage from './productImage';
import ProductDetails from './productDetails';
import DisplaySimilarProducts from './displaySimilarProducts'
// import { PromiseProvider } from 'mongoose';

export default class Product extends Component {

    state = {
        success: false,
        activePicture: '',
        gender: 'male',
        size: 'L',
        color: '',
    }

    static propType = {
        product: propType.object.isRequired,
    }

    componentDidMount() {
        this.setState({ activePicture: this.props.product.photo });

    }


    addItemToCartSuccess = (bool) => {
        this.setState({ success: bool, failed: !bool });
    }
    handleSelectPhoto = (e) => {
        const index = parseInt(e.target.attributes.photoindex.value);
        this.setState({ activePicture: this.props.product.secondaryPhotos[index] });
    }
    handleOptions = (e) => {
        if (e.target.className === 'selectGender') {
            this.setState({ gender: e.target.value });
        } else if (e.target.className === 'selectSize') {
            this.setState({ size: e.target.value });
        } else if (e.target.classList.contains('colorOption')) {
            if (e.target.id === this.state.color) return this.setState({ color: '' });
            else this.setState({ color: e.target.id });
        };
    }

    render() {
        let { props } = this;
        let { color, size, gender, activePicture } = this.state;

        return (
            <div className='productSubContainer' >
                <div className='TopSection' >

                    <ProductImage
                        activePicture={activePicture}
                        handleSelectPhoto={this.handleSelectPhoto}
                        {...props} />

                    <ProductDetails
                        handleOptions={this.handleOptions}
                        addItemToCartSuccess={this.addItemToCartSuccess}
                        {...props}
                        {...this.state}
                    />
                </div>

                <div className='BottomSection'>
                    <DisplaySimilarProducts {...props} />
                </div>
            </div>
        )
    }
}