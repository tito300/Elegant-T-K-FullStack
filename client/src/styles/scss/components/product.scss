

.product-display {
  @media all and (max-width: 850px) {
    width: 100% !important;
  }
}
  
.productSubContainer {
    display: flex;
    flex-direction: column;
    width: 100%;
    margin-bottom: 120px;
    // background-color: #efefef;
  

 .TopSection {
    height: 800px;
    // background-color: red;
    margin-bottom: 20px;
    display: flex;
    flex-direction: row;
    justify-content: center;
    padding-top: 70px;
    margin-top: -50px;
    border-top: 1px solid c(border);

    @media all and (max-width: 850px) {
      flex-direction: column;
      height: fit-content;
      justify-content: center;
    }

  };

 .BottomSection {
    height: 260px;
    margin-top: 20px;
    // background-color: green;
    // border: 2px solid green;
    padding: 0 0 25px 0;
    display: flex;
    flex-direction: column;
    align-content: center;
    overflow: hidden;
    border-top: 1px solid rgb(238, 238, 238);
    border-bottom: 1px solid rgb(238, 238, 238);

    .similarItems {
      @include font-style(arial, 30px, #515151, 600);
      margin: 20px auto 20px auto;
      width: fit-content;
    }
    .bottomGallery {
      height: 200px;
      display: flex;
      justify-content: center;
      overflow: hidden;
      align-items: center;
    }

    .bottomPhotosGallery {
      width: 900px;
      display: flex;
      justify-content: center;
      align-items: center;

    }
    .bottomItem {
      width: 130px;
      // height: 150px;
      margin-left: 35px;
      max-height: 150px;
      max-height: 130px;
      overflow: hidden;
      &:first-child {
        margin-left: 0;
      }
      img {
        opacity: 0.75;
        &:hover {
          opacity: 0.9;
        }
      }
      
    }

  };

 .photoSection {
    width: 38%;
    overFlow: hidden;
    background-color: white;
    margin-right: 10px;

    @media all and (max-width: 850px) {
      width: 100%;
    }
  };
  .photoContainer {
    position: relative;
    height: 500px;
    overflow: hidden;
    margin-bottom: 15px;
    text-align: center;
    @media all and (max-width: 850px) {
      height: fit-content;
    }
  }
 .photo {
    width: 100%;
    height: auto;
    border-radius: 2px;
    opacity: 0.95;
    transition: all 0.5s;
    &:hover {
      opacity: 1;
    }
  };
  .photo-error {
    height: 460px;
    width: 100%;
    background: #f8f8f8;
    @include font-style(futura, 30px, light, 600);
    p {
      position: absolute;
      top: 200px;
      left: 0;
      bottom: 0;
      right: 0;
    }
  }
 .gallerySection {
    display: flex;
    justify-content: center;
    flex-direction: row;
    padding-top: 5px;
  };
 .galleryContainer {
    display: flex;
    justify-content: center;
    flex-direction: row;
  };
 .imgContainer {
    width: 70px;
    height: 80px;
    margin: 0 10px;
    &.hide {
      display: none;
    }
  };
 .galleryPhoto {
    border-radius: 3px;
    width: 100%;
    opacity: 0.7;
    transition: all 0.5s;
    cursor: pointer;
    &:hover {
      transform: scale(1.1);
      opacity: 1;
    }
  };
 .fa.fa-angle-left, .fa.fa-angle-right {
   cursor: pointer;
   font-size: 60px;
   opacity: 0.8;
   margin: 0 15px;
   color: rgb(214, 214, 214);
   transition: all 0.5s;
   &:hover {
    transform: scale(1.1);
    color: rgb(199, 199, 199);
   }
 } 


.infoSection {
    width: 58%;
    overFlow: hidden;
    background-color: white;
    height: 100%;
    padding: 10px 25px;

    @media all and (max-width: 850px) {
      width: 100%;
    }

    .add-btn {
      background-color: green;
      &:hover {
        background-color: #bf3030;
      }
    }
  };
 .title {
    color: #515151;
    font-size: 37px;
    font-family: futura;
    margin-bottom: 4px;
    padding-bottom: 5px;
    border-bottom: 1px solid #c7c7c7;
  };
 .brand {
    color: #999999;
    font-size: 14px;
    font-family: arial;
    margin-bottom: 25px;
  };

 .priceContainer {
    margin-bottom: 50px;
    min-height: 100px;
    display: flex;
    justify-content: space-between;
  };
 .price {
    color: #515151;
    font-size: 15px;
    font-family: arial;
  };
 .priceNumber {
    color: #c62323;
    font-weight: 600;
    font-size: 18px;
  };
 .discreptionContainer {

  };
 .discreptionText {
    color: #515151;
    font-family: arial;
    /* padding: 10px 10px; */
  };
 .discreptionTitle {
    color: rgb(98, 98, 98);
    font-family: arial;
    margin-bottom: 5px;
    font-weight: 600;
  };

  .optionsContainer {
    margin: 30px 0;
    display: flex;
    flex-direction: column;
    align-content: flex-start;
    // height: 400px; // temporary
    & > div {
      width: 100%;
      margin: 15px 0;
    }

    .optionsTitle {
      @include font-style(arial, 18px, text-mid-dark, 600);
      margin-bottom: 10px;
    }
    .colorsContainer {
      display: flex;
      justify-content: flex-start;
      & > span {
        width: 35px;
        height: 35px;
        margin-right: 10px;
        border-radius: 5px;
        opacity: 0.7;
        cursor: pointer;
        transition: all 0.3s linear;
        &:hover {
          opacity: 0.9;
        }
        &.selectedColor {
          opacity: 0.9;
        }
      };

    };
      .selectSize {
        padding: 0 20px 0 0;
        opacity: 0.6;      
      }
      .selectGender {
        padding: 0 10px 0 0;
        opacity: 0.6; 
      } 

  };
  


};