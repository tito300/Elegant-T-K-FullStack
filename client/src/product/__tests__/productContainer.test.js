import React from 'react';
import { shallow } from 'enzyme';
import ProductContainer from '../productContainer'
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import product from '../../__mocks__/product.json'

let wrapper;

describe('ProductContainer component', ()=> {
    beforeAll(async ()=>{
        const mock = new MockAdapter(axios);
        mock.onGet('/api/products/1234')
            .reply(200, product);
        mock.onGet('/api/products/falsenumber')
            .reply(404);    

        let computedMatch = {
            params: {
                id: '1234'
            }
        }
        wrapper = await shallow(<ProductContainer computedMatch={computedMatch} />);
    })
    it('should match snapshot', ()=> {
        expect(wrapper).toMatchSnapshot();
    });
    it('should display spinner when fetching products', ()=>{
        wrapper.setState({
            product: null,
            pending: true,
            err: false,
            errMsg: null,
        })

        expect(wrapper.find('.fa-spinner').exists()).toBeTruthy();
    });
    it('should show a 404 when product is not available',async ()=>{
        let computedMatch = {
            params: {
                id: 'falsenumber'
            }
        }
        let wrapper2 = await shallow(<ProductContainer computedMatch={computedMatch} />);
        await wrapper2.instance().componentDidMount();
        // console.log(wrapper2.state());
        // console.log(wrapper2.props());
        // console.log(wrapper2.html());

        // expect(wrapper2.find('.product-list__Error').exists()).toBeTruthy();

    })
})