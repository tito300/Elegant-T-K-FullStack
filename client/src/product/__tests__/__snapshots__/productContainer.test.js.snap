// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductContainer component should match snapshot 1`] = `
<div
  className="product-display"
  style={
    Object {
      "margin": "150px auto 0",
      "minHeight": "calc(100vh - 78px)",
      "overFlow": "hidden",
      "paddingTop": "30px",
      "width": "1200px",
    }
  }
>
  <ErrorBoundary>
    <Product
      product={
        Object {
          "discreption": "random description",
          "id": "1234",
          "lastModified": "2019-01-10T05:19:59.000Z",
          "material": "fiber",
          "photo": "imgs/1.jpg",
          "price": "20.00",
          "primaryColor": "red",
          "purchaseCount": 6,
          "rating": 3,
          "secondaryPhotos": Array [
            Object {
              "color": "black",
              "link": "imgs/1-1.jpg",
            },
            Object {
              "color": "grey",
              "link": "imgs/1-2.jpg",
            },
          ],
          "title": "random product",
          "uploadDate": "2018-08-07T20:37:06.682Z",
          "variants": Object {
            "female": Array [
              Object {
                "color": "black",
                "sizes": Array [
                  Object {
                    "color": "black",
                    "gender": "female",
                    "photo": "imgs/1.jpg",
                    "price": "21.00",
                    "size": "l",
                  },
                  Object {
                    "color": "black",
                    "gender": "female",
                    "photo": "imgs/1.jpg",
                    "price": "231.00",
                    "size": "s",
                  },
                ],
              },
            ],
            "male": Array [
              Object {
                "color": "red",
                "sizes": Array [
                  Object {
                    "color": "red",
                    "gender": "male",
                    "photo": "imgs/1.jpg",
                    "price": "23.00",
                    "size": "l",
                  },
                  Object {
                    "color": "red",
                    "gender": "male",
                    "photo": "imgs/1.jpg",
                    "price": "23.00",
                    "size": "s",
                  },
                ],
              },
            ],
          },
          "weight": "5",
        }
      }
    />
  </ErrorBoundary>
</div>
`;
